x-pack/metricbeat/module/aws/sqs/sqs.go

```diff
@@ -0,0 +1,194 @@
+		mb.DefaultMetricSet(),
+	)
+}
+
+// MetricSet holds any configuration or state information. It must implement
+// the mb.MetricSet interface. And this is best achieved by embedding
+// mb.BaseMetricSet because it implements all of the required mb.MetricSet
+// interface methods except for Fetch.
+type MetricSet struct {
+	*aws.MetricSet
+	logger *logp.Logger
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	logger := logp.NewLogger(aws.ModuleName)
+
+	moduleConfig := aws.Config{}
+	if err := base.Module().UnpackConfig(&moduleConfig); err != nil {
+		return nil, err
+	}
+
+	if moduleConfig.Period == "" {
+		err := errors.New("period is not set in AWS module config")
+		logger.Error(err)
+	}
+
+	metricSet, err := aws.NewMetricSet(base)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating aws metricset")
+	}
+
+	// Check if period is set to be multiple of 300s
+	remainder := metricSet.PeriodInSec % 300
+	if remainder != 0 {
+		err := errors.New("period needs to be set to 300s (or a multiple of 300s). " +
+			"To avoid data missing or extra costs, please make sure period is set correctly in config.yml")
+		logger.Info(err)
+	}
+
+	return &MetricSet{
+		MetricSet: metricSet,
+		logger:    logger,
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(report mb.ReporterV2) {
+	namespace := "AWS/SQS"
+	// Get startTime and endTime
+	startTime, endTime, err := aws.GetStartTimeEndTime(m.DurationString)
+	if err != nil {
+		logp.Error(errors.Wrap(err, "Error ParseDuration"))
+		m.logger.Error(err.Error())
+		report.Error(err)
+		return
+	}
+
+	for _, regionName := range m.MetricSet.RegionsList {
+		m.MetricSet.AwsConfig.Region = regionName
+		svcCloudwatch := cloudwatch.New(*m.MetricSet.AwsConfig)
+		listMetricsOutput, err := aws.GetListMetricsOutput(namespace, regionName, svcCloudwatch)
+		if err != nil {
+			m.logger.Error(err.Error())
+			report.Error(err)
+			continue
+		}
+
+		if listMetricsOutput == nil || len(listMetricsOutput) == 0 {
+			continue
+		}
+
+		metricDataQueries := constructMetricQueries(listMetricsOutput, int64(m.PeriodInSec))
+		if len(metricDataQueries) == 0 {
+			continue
+		}
+
+		init := true
+		getMetricDataOutput := &cloudwatch.GetMetricDataOutput{NextToken: nil}
+		for init || getMetricDataOutput.NextToken != nil {
+			init = false
+			output, err := aws.GetMetricDataPerRegion(metricDataQueries, getMetricDataOutput.NextToken, svcCloudwatch, startTime, endTime)
+			if err != nil {
+				err = errors.Wrap(err, "getMetricDataPerRegion failed, skipping region "+regionName)
```

Comment:
I see this pattern across all our metricsets, we should come up wit a 1 liner to do this. @jsoriano FYI 

I think @ycombinator already did something about this in the ES module.